### PR TITLE
[core] Report resource load by shape

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

This is an experimental feature to support better autoscaling. This PR reports the resource load at each node based on how many requests are queued per resource shape, instead of an aggregate of all of the resource requests queued. This should allow for smarter bin packing by the autoscaler.

The load includes stats from the ready and infeasible queues. The counts are piggy-backed onto the heartbeat messages to the GCS, which aggregates them and publishes the load to the autoscaler.

The feature can be enabled by setting `max_resource_shapes_per_load_report` in the internal config to nonzero. There is also an option called `num_heartbeats_between_load_reports` to adjust how often load reports are sent.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_node_manager.cc [hunk 1]
```diff
@@ -89,8 +89,34 @@ void GcsNodeManager::NodeFailureDetector::DetectDeadNodes() {
...
     auto batch = std::make_shared<rpc::HeartbeatBatchTableData>();
-    for (const auto &heartbeat : heartbeat_buffer_) {
-      batch->add_batch()->CopyFrom(heartbeat.second);
+    std::unordered_map<SchedulingClass, rpc::ResourceDemand> aggregate_load;
+    for (auto &heartbeat : heartbeat_buffer_) {
+      // Aggregate the load reported by each raylet.
+      auto load = heartbeat.second.resource_load_by_shape();
+      for (const auto &demand : load.resource_demands()) {
+        auto scheduling_class =
+            TaskSpecification::GetSchedulingClass(MapFromProtobuf(demand.shape()));
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_node_manager.cc:98]: It's a little odd to use the scheduling class a temporary key here, I guess there isn't any other convenient key function that would work?

File: src/ray/protobuf/gcs.proto [hunk 2]
```diff
@@ -253,6 +253,16 @@ message GcsNodeInfo {
   string node_manager_hostname = 8;
 }
 
+message ResourceDemand {
+  map<string, double> shape = 1;
+  uint64 num_ready_requests_queued = 2;
+  uint64 num_infeasible_requests_queued = 3;
```
Comment (ericl) [inline src/ray/protobuf/gcs.proto:259]: Can you document these?

File: src/ray/protobuf/gcs.proto [hunk 3]
```diff
@@ -262,14 +272,16 @@ message HeartbeatTableData {
   map<string, double> resources_total = 3;
   // Aggregate outstanding resource load on this node manager.
   map<string, double> resource_load = 4;
+  ResourceLoad resource_load_by_shape = 5;
```
Comment (ericl) [inline src/ray/protobuf/gcs.proto:275]: Docstring

File: src/ray/raylet/node_manager.cc [hunk 4]
```diff
@@ -457,6 +459,17 @@ void NodeManager::Heartbeat() {
         ResourceSet(local_resources.GetLoadResources()));
   }
 
+  // Add resource load by shape. This will be used by the new autoscaler.
+  if (num_heartbeats_before_load_report_ == 0) {
```
Comment (ericl) [inline src/ray/raylet/node_manager.cc:463]: Doesn't this make aggregation difficult / incorrect if some heartbeats are missing the data? I would prefer to leave this out in the initial version. Also, Ant's delta heartbeat updates might already provide a similar bandwidth saving affect.

File: src/ray/common/ray_config_def.h [hunk 5]
```diff
@@ -333,3 +333,13 @@ RAY_CONFIG(int, metrics_agent_port, -1)
 /// the owner has been granted a lease. A value >1 is used when we want to enable
 /// pipelining task submission.
 RAY_CONFIG(uint32_t, max_tasks_in_flight_per_worker, 1)
+
+/// The maximum number of resource shapes included in the resource
+/// load reported by each raylet.
+RAY_CONFIG(int64_t, max_resource_shapes_per_load_report, 0)
```
Comment (ericl) [inline src/ray/common/ray_config_def.h:339]: Can we enable this by default to 100?

**Review by ericl — CHANGES_REQUESTED** [review]

File: src/ray/raylet/node_manager.cc (hunk 4 above)
Comment (stephanie-wang) [reply src/ray/raylet/node_manager.cc:463]: Ah right, I forgot that we're aggregating across multiple nodes. I'll remove it.

File: src/ray/gcs/gcs_server/gcs_node_manager.cc (hunk 1 above)
Comment (stephanie-wang) [reply src/ray/gcs/gcs_server/gcs_node_manager.cc]: I replaced this with the `ResourceSet` class.

**Review by ericl — APPROVED** [review]
LGTM

**Pull request merged**